File: architecture-skills/object-oriented-programming-paradigm/examples/composition-over-inheritance/composition-over-inheritance.swift
// Composition over inheritance in Swift.
//
// Robot holds two protocol-typed collaborators instead of subclassing
// every movement x weapon combination. Parts can be swapped at runtime.

import Foundation

protocol MoveBehavior {
    func move() -> String
}

protocol AttackBehavior {
    func attack() -> String
}

struct Treads: MoveBehavior {
    func move() -> String { "rolls forward on treads" }
}

struct Rotors: MoveBehavior {
    func move() -> String { "lifts off on quad rotors" }
}

struct Turbines: MoveBehavior {
    func move() -> String { "dives and swims with turbines" }
}

struct Laser: AttackBehavior {
    func attack() -> String { "fires a precision laser" }
}

struct Drill: AttackBehavior {
    func attack() -> String { "spins up a mining drill" }
}

final class Robot {
    let name: String
    var mover: MoveBehavior
    let weapon: AttackBehavior

    init(name: String, mover: MoveBehavior, weapon: AttackBehavior) {
        self.name = name
        self.mover = mover
        self.weapon = weapon
    }

    func perform() -> String {
        String(format: "%8@: %@ and %@", name, mover.move(), weapon.attack())
    }
}

var squad = [
    Robot(name: "Scout", mover: Treads(), weapon: Laser()),
    Robot(name: "Miner", mover: Treads(), weapon: Drill()),
    Robot(name: "Wasp", mover: Rotors(), weapon: Drill()),
    Robot(name: "Manta", mover: Turbines(), weapon: Laser()),
]

for robot in squad {
    print(robot.perform())
}

print("-- field upgrade: Miner receives rotors --")
squad[1].mover = Rotors()
print(squad[1].perform())

File: architecture-skills/object-oriented-programming-paradigm/examples/composition-over-inheritance/composition-over-inheritance.php
<?php
/**
 * Composition over inheritance in PHP.
 *
 * Robot holds MoveBehavior and AttackBehavior collaborators instead of
 * subclassing every movement x weapon combination. Parts swap at runtime.
 */

declare(strict_types=1);

interface MoveBehavior
{
    public function move(): string;
}

interface AttackBehavior
{
    public function attack(): string;
}

final class Treads implements MoveBehavior
{
    public function move(): string { return 'rolls forward on treads'; }
}

final class Rotors implements MoveBehavior
{
    public function move(): string { return 'lifts off on quad rotors'; }
}

final class Turbines implements MoveBehavior
{
    public function move(): string { return 'dives and swims with turbines'; }
}

final class Laser implements AttackBehavior
{
    public function attack(): string { return 'fires a precision laser'; }
}

final class Drill implements AttackBehavior
{
    public function attack(): string { return 'spins up a mining drill'; }
}

final class Robot
{
    public function __construct(
        private readonly string $name,
        private MoveBehavior $mover,
        private readonly AttackBehavior $weapon,
    ) {}

    public function setMover(MoveBehavior $mover): void
    {
        $this->mover = $mover;
    }

    public function perform(): string
    {
        return sprintf(
            '%8s: %s and %s',
            $this->name,
            $this->mover->move(),
            $this->weapon->attack()
        );
    }
}

$squad = [
    new Robot('Scout', new Treads(), new Laser()),
    new Robot('Miner', new Treads(), new Drill()),
    new Robot('Wasp', new Rotors(), new Drill()),
    new Robot('Manta', new Turbines(), new Laser()),
];

foreach ($squad as $robot) {
    echo $robot->perform(), PHP_EOL;
}

echo "-- field upgrade: Miner receives rotors --\n";
$squad[1]->setMover(new Rotors());
echo $squad[1]->perform(), PHP_EOL;
